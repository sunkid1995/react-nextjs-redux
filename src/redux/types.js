export const PREFIXES = {
  AUTH: 'auth/',
  SERVICE_API: 'serviceApi/',
  SHARED: 'shared/',
};

export const INIT_ACTION = { type: '@@redux/INIT' };

export const AUTH = {
  AUTHORIZE: 'auth/AUTHORIZE',
  DEAUTHORIZE: 'auth/DEAUTHORIZE',
};

export const SERVICE_API = {
  AUTH_WITH_EMAIL: 'serviceApi/AUTH_WITH_EMAIL',
  FORGOT_PASSWORD: 'serviceApi/FORGOT_PASSWORD',
  RESET_PASSWORD: 'serviceApi/RESET_PASSWORD',
};

export const SHARED = {

};
