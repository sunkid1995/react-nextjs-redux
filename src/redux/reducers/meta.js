import { REHYDRATE } from 'redux-persist/lib/constants';

const INITIAL_STATE = {
  rehydrated: false,
};

export default function metaReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === REHYDRATE) {
    const { err } = action;

    /**
     * On server-side, REHYDRATE action will throw an err:
     * Error: No available storage method found
     * Otherwise on client-side, REHYDRATE will be dispatched without error
     * So we check if err is nullable then it's actually rehydrated in client
     */
    if (err == null) return { ...state, rehydrated: true };
  }

  return state;
}
