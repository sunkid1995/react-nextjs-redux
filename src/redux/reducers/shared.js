// Redux
// import { SHARED } from 'src/redux/types';

// const INITIAL_STATE = {
//   handleShipping: undefined,
//   type: false,
// };

// export default function sharedReducer(state = INITIAL_STATE, action) {
//   const { type } = action;

//   return state;
// }
