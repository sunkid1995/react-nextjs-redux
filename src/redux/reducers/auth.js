// Redux
import { AUTH } from 'src/redux/types';

const INITIAL_STATE = {
  user: undefined,
};

export default function authReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === AUTH.AUTHORIZE) {
    const { payload: user } = action;
    return { ...state, user };
  }

  if (type === AUTH.DEAUTHORIZE) {
    return INITIAL_STATE;
  }

  return state;
}
