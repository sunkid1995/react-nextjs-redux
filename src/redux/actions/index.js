import * as auth from './auth';
import * as serviceApi from './serviceApi';
import * as shared from './shared';

export { auth, serviceApi, shared };
