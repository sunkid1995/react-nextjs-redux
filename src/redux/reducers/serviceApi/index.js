import _ from 'lodash';

// Constants
import { AUTH, SERVICE_API } from 'src/redux/types';

// Locals
// import blankRequestReducer from './blankRequestReducer';
// import paginatedRequestReducer from './paginatedRequestReducer';
import plainRequestReducer from './plainRequestReducer';

const INITIAL_BLANK_REQUEST = { error: undefined, loading: false };
// const INITIAL_PAGINATED_DATA_REQUEST = { dataPages: {}, totalPages: 1, ...INITIAL_BLANK_REQUEST };
const INITIAL_PLAIN_DATA_REQUEST = { data: {}, ...INITIAL_BLANK_REQUEST };

const INITIAL_STATE = {
  emailAuth: INITIAL_PLAIN_DATA_REQUEST,
  passwordForgotten: INITIAL_PLAIN_DATA_REQUEST,
  passwordReset: INITIAL_PLAIN_DATA_REQUEST,
};

export default function serviceApiReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === AUTH.DEAUTHORIZE) return INITIAL_STATE;

  if (
    _.startsWith(type, SERVICE_API.AUTH_WITH_EMAIL) ||
    _.startsWith(type, SERVICE_API.FORGOT_PASSWORD) ||
    _.startsWith(type, SERVICE_API.RESET_PASSWORD)
  ) return plainRequestReducer(state, action);

  return state;
}
