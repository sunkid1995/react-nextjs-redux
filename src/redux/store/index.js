import { applyMiddleware, compose, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import { persistCombineReducers, persistStore } from 'redux-persist';
import reduxThunk from 'redux-thunk';

import { IN_DEV_MODE } from 'src/constants';
import { auth, meta, serviceApi, shared } from 'src/redux/reducers';

// Locals
import buildAxiosMiddleware from './buildAxiosMiddleware';
import buildPersistorConfigs from './buildPersistorConfigs';

const composeEnhancer = IN_DEV_MODE ? composeWithDevTools({}) : compose;

export default function makeStore(/* initialState, options */) {
  const configs = buildPersistorConfigs();

  const store = createStore(
    persistCombineReducers(configs, { auth, meta, serviceApi, shared }),
    composeEnhancer(applyMiddleware(buildAxiosMiddleware(), reduxThunk)),
  );

  persistStore(store);
  return store;
}
