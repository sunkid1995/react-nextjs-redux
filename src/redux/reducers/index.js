import auth from './auth';
import meta from './meta';
import serviceApi from './serviceApi';
import shared from './shared';

export { auth, meta, serviceApi, shared };
