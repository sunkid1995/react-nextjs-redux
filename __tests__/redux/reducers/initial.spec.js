import { auth, meta, serviceApi } from 'src/redux/reducers';

// Helpers
import buildMockState from '__tests__/helpers/buildMockState';

describe('redux/reducers/initial', () => {
  const mockState = buildMockState();

  test('auth', () => {
    const { auth: expectedState } = mockState;

    expect(auth(undefined, { type: 'DUMMY_TYPE' })).toEqual(expectedState);
    expect(auth(undefined, { type: 'DUMMY_TYPE' })).toMatchSnapshot();
  });

  test('meta', () => {
    const { meta: expectedState } = mockState;

    expect(meta(undefined, { type: 'DUMMY_TYPE' })).toEqual(expectedState);
    expect(meta(undefined, { type: 'DUMMY_TYPE' })).toMatchSnapshot();
  });

  test('serviceApi', () => {
    const { serviceApi: expectedState } = mockState;

    expect(serviceApi(undefined, { type: 'DUMMY_TYPE' })).toEqual(expectedState);
    expect(serviceApi(undefined, { type: 'DUMMY_TYPE' })).toMatchSnapshot();
  });
});
