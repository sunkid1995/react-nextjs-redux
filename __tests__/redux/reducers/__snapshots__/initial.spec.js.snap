// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux/reducers/initial auth 1`] = `
Object {
  "user": undefined,
}
`;

exports[`redux/reducers/initial meta 1`] = `
Object {
  "rehydrated": false,
}
`;

exports[`redux/reducers/initial serviceApi 1`] = `
Object {
  "emailAuth": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
  "passwordForgotten": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
  "passwordReset": Object {
    "data": undefined,
    "error": undefined,
    "loading": false,
  },
}
`;
